dance/automl_config/step2_config.py

```diff
@@ -0,0 +1,143 @@
+import functools
+import itertools
+
+import wandb
+
+from dance.automl_config.fun2code import fun2code_dict
+from dance.transforms.misc import SetConfig
+
+#TODO register more functions and add more examples
+pipline2fun_dict = {
+    "normalize": {
+        "values": ["normalize_total", "log1p", "scaleFeature", "scTransform"]
+    },
+    "gene_filter": {
+        "values":
+        ["filter_gene_by_count", "filter_gene_by_percentile", "highly_variable_genes", "Filter_gene_by_regress_score"]
+    },
+    "gene_dim_reduction": {
+        "values": ["cell_svd", "cell_weighted_pca", "cell_pca"]
+    },
+    "cell_filter": {
+        "values": ["filter_cell_by_count"]
+    },
+    "mask_name": {
+        "values": ["cell_wise_mask_data", "mask_data"]
+    },
+    "gene_hold_out_name": {
+        "values": ["gene_hold_out"]
+    }
+}  #Functions registered in the preprocessing process
+
+
+def generate_combinations_with_required_elements(elements, required_elements=[]):
+    optional_elements = [x for x in elements if x not in required_elements]
+
+    # Sort optional elements in the same order as in the `elements` list
+    optional_elements.sort(key=lambda x: elements.index(x))
+
+    # Generate all possible combinations of optional elements
+    optional_combinations = []
+    for i in range(len(optional_elements) + 1):
+        optional_combinations += list(itertools.combinations(optional_elements, i))
+
+    # Combine required elements with optional combinations to get all possible combinations
+    all_combinations = []
+    for optional_combination in optional_combinations:
+        all_combinations.append([x for x in elements if x in required_elements or x in optional_combination])
+    return all_combinations
+
+
+def getFunConfig(selected_keys=None):
+    """Get the config that needs to be optimized and the number of rounds."""
+    global pipline2fun_dict
+    pipline2fun_dict_subset = {key: pipline2fun_dict[key] for key in selected_keys}
+    print(pipline2fun_dict)
+    count = 1
+    for _, pipline_values in pipline2fun_dict_subset.items():
+        count *= len(pipline_values['values'])
+    return pipline2fun_dict_subset, count
+
+
+def get_transforms(config=None, set_data_config=True, save_raw=False):
+    """Obtain the Compose of the preprocessing function according to the preprocessing
+    process."""
+    if ("normalize" not in config.keys() or config.normalize
+            != "log1p") and ("gene_filter" in config.keys() and config.gene_filter == "highly_variable_genes"):
+
+        return None
```

Comment:
When transform is None, it represents an exception, but I don't want to stop searching